llvm/lib/CodeGen/MachineTraceMetrics.cpp

```diff
@@ -761,6 +762,64 @@ static void updatePhysDepsDownwards(const MachineInstr *UseMI,
   }
 }
 
+/// Estimates the number of cycles elapsed between DefMI and UseMI if they're
+/// non-null and in the same BasicBlock. Returns std::nullopt when UseMI is in a
+/// different MBB than DefMI.
+static std::optional<unsigned>
+estimateDefUseCycles(const TargetSchedModel &Sched, const MachineInstr *DefMI,
+                     const MachineInstr *UseMI) {
+  if (!DefMI || !UseMI || DefMI == UseMI)
+    return 0;
+  const MachineBasicBlock *ParentBB = DefMI->getParent();
+  if (ParentBB != UseMI->getParent())
+    return std::nullopt;
+
+  const auto DefIt =
+      llvm::find_if(ParentBB->instrs(),
+                    [DefMI](const MachineInstr &MI) { return DefMI == &MI; });
+  const auto UseIt =
+      llvm::find_if(ParentBB->instrs(),
+                    [UseMI](const MachineInstr &MI) { return UseMI == &MI; });
+
+  unsigned NumMicroOps = 0;
+  for (auto It = DefIt; It != UseIt; ++It) {
+    // In cases where the UseMI is a PHI at the beginning of the MBB, compute
+    // MicroOps until the end of the MBB.
+    if (It.isEnd())
+      break;
+
+    NumMicroOps += Sched.getNumMicroOps(&*It);
+  }
+  return NumMicroOps / Sched.getIssueWidth();
+}
+
+/// Wraps Sched.computeOperandLatency, accounting for the case when
+/// InstrSchedModel and InstrItineraries are not available: in this case,
+/// Sched.computeOperandLatency returns DefaultDefLatency, which is a very rough
+/// approximate; to improve this approximate, offset it by the approximate
+/// cycles elapsed from DefMI to UseMI (since the MIs could be re-ordered by the
+/// scheduler, and we don't have this information, this cannot be known
+/// exactly). When scheduling information is available,
```

Comment:
If we know this is an OOO core, does this change really make sense since instructions will be reordered? Would it be better to use the default def latency in those cases?

Reply:
From a software (compiler) perspective, isn't it impossible to know how an OOO core will re-order instructions? Independent of this patch, we have a scheduling model that we use that cannot know about the exact re-ordering (please correct me if I got something wrong). IIUC, a scheduler descriptor simply states whether a core is in-order or OOO.

Reply:
The argument being made in this patch is that order of instructions should be used to estimate latency. But on an OOO core, we don't know the order. So for that reason I am asking whether using order when we are OOO is preferred over using latency.

Reply:
Okay, this is a subtle point, to which I don't have a good answer. The scheduler can also re-order instructions, depending on scheduling information (which also contains information about in-order/OOO). I just have some practical answers, and would be happy if someone else has better answers.

1. Do we know whether the core is in-order/OOO without the scheduling model?
2. I did the benchmarking on OOO cores. We got positive test changes and speedups.